In [2]:
import numpy as np

2.4 Compute the matrix products:

\begin{gathered}
\begin{bmatrix}
1& 2& 3 \\
4& 5& 6 \\
7& 8& 9  
\end{bmatrix} 

\begin{bmatrix}
1& 1& 1 \\
0& 1& 1 \\
1& 0& 1  
\end{bmatrix}

\end{gathered}

In [6]:
a = np.array([
    [1,2,3],
    [4,5,6],
    [7,8,9]
])

b = np.array([
    [1,1,0],
    [0,1,1],
    [1,0,1]
])

print(np.dot(a,b))

[[ 4  3  5]
 [10  9 11]
 [16 15 17]]


\begin{gathered}
\begin{bmatrix}
1 & 1 & 0 \\
0&  1&1  \\
1&  0&  1
\end{bmatrix} 
\begin{bmatrix}
1&2&3  \\
4&5&6  \\
7&8&9  
\end{bmatrix}
\end{gathered}

In [7]:
print(b @ a)

[[ 5  7  9]
 [11 13 15]
 [ 8 10 12]]


\begin{gathered}
\begin{bmatrix}
1 & 2 & 1&2 \\
4&1&-1&-4
\end{bmatrix} 
\begin{bmatrix}
0&3  \\
1&-1  \\
2&1 \\
5&2  
\end{bmatrix}
\end{gathered}

In [10]:
a = np.array([
    [1,2,1,2],
    [4,1,-1,-4]
])

b = np.array([
    [0,3],
    [1,-1],
    [2,-1],
    [5,2]
])

print(a @ b)

[[ 14   4]
 [-21   4]]


\begin{gathered}
\begin{bmatrix}
0&3  \\
1&-1  \\
2&1 \\
5&2  
\end{bmatrix}
\begin{bmatrix}
1 & 2 & 1&2 \\
4&1&-1&-4
\end{bmatrix} 
\end{gathered}

In [11]:
print(b @ a)

[[ 12   3  -3 -12]
 [ -3   1   2   6]
 [ -2   3   3   8]
 [ 13  12   3   2]]


2.5 Find the set S of all solutions in x of the inhomogeneous linear equation systems

#### $$Ax = b $$

Then, $$ A^{-1}b = x $$

 a)
\begin{gathered} A = 
\begin{bmatrix}
1&1&-1&-1  \\
2&5&-7&-5  \\
2&-1&1&3 \\
5&2&-4&2
\end{bmatrix} ,
b =\begin{bmatrix}
1 \\
-2 \\
4 \\
6
\end{bmatrix} 
\end{gathered}

In [14]:
A = np.array([
    [1,1,-1,-1],
    [2,5,-7,5],
    [2,-1,1,3],
    [5,2,-4,2]
])

b = np.array([1,-2,4,6])

In [15]:
print(np.linalg.solve(A,b))

[ 1.73333333 -0.43333333  0.4        -0.1       ]


Then : 
$$ x = \begin{bmatrix}
1.73 \\
-4.33 \\
0.4 \\
-0.1 \\
\end{bmatrix} $$

b)

\begin{gathered} A = 
\begin{bmatrix}
1&-1&0&0&1  \\
1&1&0&-3&0  \\
2&-1&0&1&-1 \\
-1&2&0&-2&-1
\end{bmatrix} ,



b =\begin{bmatrix}
3 \\
6 \\
5 \\
1
\end{bmatrix} 
\end{gathered}

In [20]:
A = np.array([
    [1,-1,0,0,1],
    [1,1,0,-3,0],
    [2,-1,0,1,-1],
    [-1,2,0,-2,-1]
])

b = np.array([3,6,5,1])

In [23]:
a_inv = np.linalg.inv(A)

LinAlgError: Last 2 dimensions of the array must be square